File: src/main/java/com/better/alarm/Store.java
package com.better.alarm;

import com.better.alarm.model.AlarmValue;
import com.google.common.base.Optional;

import org.immutables.value.Value;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.Subject;

/**
 * Created by Yuriy on 10.06.2017.
 */

@Value.Immutable
@Value.Style(stagedBuilder = true)
public abstract class Store {
    @Value.Immutable
    @Value.Style(stagedBuilder = true)
    public interface Next {

        AlarmValue alarm();

        boolean isPrealarm();

        long nextNonPrealarmTime();
    }

    @Value.Immutable
    public interface AlarmSet {
        @Value.Parameter
        AlarmValue alarm();

        @Value.Parameter
        long millis();
    }

    public Observable<List<AlarmValue>> alarms(){
        return alarmsSubject().distinctUntilChanged();
    }

    public abstract BehaviorSubject<List<AlarmValue>> alarmsSubject();

    public abstract BehaviorSubject<Optional<Next>> next();

    public abstract Subject<AlarmSet> sets();
}

File: app/src/main/java/com/better/alarm/domain/Store.kt
package com.better.alarm.domain

import com.better.alarm.data.AlarmValue
import com.better.alarm.services.Event
import com.better.alarm.util.Optional
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.Subject

/** Created by Yuriy on 10.06.2017. */
data class Store(
    val alarmsSubject: BehaviorSubject<List<AlarmValue>>,
    val next: BehaviorSubject<Optional<Next>>,
    val sets: Subject<AlarmSet>,
    val events: Subject<Event>,
) {
  val uiVisible: BehaviorSubject<Boolean> = BehaviorSubject.createDefault(false)

  fun alarms(): Observable<List<AlarmValue>> {
    return alarmsSubject().distinctUntilChanged()
  }

  fun alarmsSubject(): BehaviorSubject<List<AlarmValue>> = alarmsSubject

  fun next(): BehaviorSubject<Optional<Next>> = next

  fun sets(): Subject<AlarmSet> = sets

  data class Next(
      val isPrealarm: Boolean,
      val alarm: AlarmValue,
      val nextNonPrealarmTime: Long,
  ) {
    fun alarm(): AlarmValue = alarm

    fun nextNonPrealarmTime(): Long = nextNonPrealarmTime
  }

  data class AlarmSet(
      val alarm: AlarmValue,
      val millis: Long,
      val fromUserInteraction: Boolean,
  ) {
    fun alarm(): AlarmValue = alarm

    fun millis(): Long = millis
  }
}
